src/object/cli_ec.c

```diff
@@ -2610,17 +2620,35 @@ struct oes_copy_arg {
 	void		*buf;
 	uint64_t	 size;
 	uint64_t	 copied;
+	d_sg_list_t	*sgl;
+	struct daos_sgl_idx *sgl_idx;
 };
 
 static int
 oes_copy(uint8_t *buf, size_t len, void *data)
 {
 	struct oes_copy_arg	*arg = data;
+	d_sg_list_t		*sgl = arg->sgl;
+	struct daos_sgl_idx	*idx = arg->sgl_idx;
 
 	D_ASSERT(arg->copied + len <= arg->size);
 	memcpy(buf, arg->buf + arg->copied, len);
 	arg->copied += len;
 
+	if (idx->iov_offset == 0) {
+		D_ASSERT(idx->iov_idx > 0);
+		sgl->sg_iovs[idx->iov_idx - 1].iov_len =
+		     sgl->sg_iovs[idx->iov_idx - 1].iov_buf_len;
+	} else {
+		sgl->sg_iovs[idx->iov_idx].iov_len =
+			max(sgl->sg_iovs[idx->iov_idx].iov_len, idx->iov_offset);
+		D_ASSERTF(sgl->sg_iovs[idx->iov_idx].iov_len <=
+			  sgl->sg_iovs[idx->iov_idx].iov_buf_len,
+			  "iov_idx %u %p/%p offset %zd + len %zd > buf_len %zd",
+			  idx->iov_idx, sgl, sgl->sg_iovs[idx->iov_idx].iov_buf,
+			  idx->iov_offset, len, sgl->sg_iovs[idx->iov_idx].iov_buf_len);
```

Comment:
this code looks confuse that why must needed, because at the end the obj_ec_fetch_set_sgl() -> dc_sgl_out_set() will set each sg_iovs[]'s iov_len and sgl->sg_nr_out. 
do you know why need such extra code here? or any case that may miss the calling of obj_ec_fetch_set_sgl()?

not sure if can with a reproducer for the problem that may be helpful to know what's the exact issue...

Reply:
So during EC fetch,  each shard fetch the data_size from the server, then we use these data_sizes from different shard to reset the input sql buffer in obj_ec_fetch_set_sgl().

But if one shard is missing, then the fetch is redirected to the parity shard, and on the parity shard, the "data size" actually depends on the partial data on the replicate space, not the real data size, i.e. the "data size" can not be trusted in obj_ec_fetch_set_sgl(). So in this case, we will set the buffer size in obj_ec_recov_data()->obj_ec_recov_fill_back()->oes_copy(),  instead of obj_ec_fetch_set_sgl().